src/main/java/graphql/execution/ExecutionStrategy.java

```diff
@@ -494,19 +496,19 @@ protected FieldValueInfo completeValueForList(ExecutionContext executionContext,
      */
     protected FieldValueInfo completeValueForList(ExecutionContext executionContext, ExecutionStrategyParameters parameters, Iterable<Object> iterableValues) {
 
-        Collection<Object> values = FpKit.toCollection(iterableValues);
+        OptionalInt size = FpKit.toSize(iterableValues);
         ExecutionStepInfo executionStepInfo = parameters.getExecutionStepInfo();
 
-        InstrumentationFieldCompleteParameters instrumentationParams = new InstrumentationFieldCompleteParameters(executionContext, parameters, () -> executionStepInfo, values);
+        InstrumentationFieldCompleteParameters instrumentationParams = new InstrumentationFieldCompleteParameters(executionContext, parameters, () -> executionStepInfo, iterableValues);
         Instrumentation instrumentation = executionContext.getInstrumentation();
 
         InstrumentationContext<ExecutionResult> completeListCtx = instrumentation.beginFieldListComplete(
                 instrumentationParams
         );
 
-        List<FieldValueInfo> fieldValueInfos = new ArrayList<>(values.size());
+        List<FieldValueInfo> fieldValueInfos = new ArrayList<>(size.orElse(1));
```

Comment:
At the end of the day we are still allocated an array the size of the list -  So source of the data can be a stream but we can stay streaming all the way.

This is an improvement - as you say it greatly reduces your memory usage (nearly half??) but longer term could we have streams all the way down to the turtles?

Thinking aloud here it might be too theoretical - I know there is code out there that assumes ExecutionResult is really a Map of lists and maps and values.  Would introducing "streams" in that break things?

Reply:
> At the end of the day we are still allocated an array the size of the list - So source of the data can be a stream but we can stay streaming all the way.

yes, I was not able to remove this big list. 

> This is an improvement - as you say it greatly reduces your memory usage (nearly half??) but longer term could we have streams all the way down to the turtles?
Don't know yet: eventually everything is "materialized" in a single ExecutionResult. Would be nice to minimize everything else. 

This the most big queries we have so far:

Before
![image](https://user-images.githubusercontent.com/5798698/95196251-1b764c00-07d8-11eb-9861-926c2e5578b2.png)

After
![image](https://user-images.githubusercontent.com/5798698/95196344-3052df80-07d8-11eb-9332-ae013af87aaf.png)

> Thinking aloud here it might be too theoretical - I know there is code out there that assumes ExecutionResult is really a Map of lists and maps and values. Would introducing "streams" in that break things?
We are merely introducing streaming between datafetchers and execution strategy: so this is very conservative choice by now.